source/ZipFile.cpp

```diff
@@ -0,0 +1,151 @@
+/* ZipFile.cpp
+Copyright (c) 2025 by tibetiroka
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "ZipFile.h"
+
+#include "Files.h"
+
+#include <functional>
+#include <numeric>
+
+using namespace std;
+
+
+
+ZipFile::ZipFile(const filesystem::path &zipPath)
+	: basePath(zipPath)
+{
+	lock_guard<recursive_mutex> guard(lock);
+	zipFile = unzOpen(basePath.string().c_str());
+	if(!zipFile)
+		throw runtime_error("Failed to open ZIP file" + zipPath.generic_string());
+
+	// Check whether this zip has a single top-level directory (such as high-dpi.zip/high-dpi)
+	filesystem::path topLevel;
+	for(const filesystem::path &path : ListFiles("", true, false))
+	{
+		filesystem::path zipPath = path.lexically_relative(basePath);
+		if(topLevel.empty())
+			topLevel = *zipPath.begin();
+		else if(*zipPath.begin() != topLevel)
+			return;
+	}
+	topLevelDirectory = topLevel;
+}
+
+
+
+ZipFile::~ZipFile()
+{
+	if(zipFile)
+	{
+		lock_guard<recursive_mutex> guard(lock);
+		unzClose(zipFile);
```

Comment:
Does zipFile need to be set to nullptr after closing?

Reply:
Nah. We have to close it because it doesn't get closed upon destruction (it's a pointer type), which means we don't have to set it to anything here. This is the destructor, so no other operations would try to access that handle.

Reply:
Agreed, but as mentioned above; then we should also not need to lock here. The access layer above ZipFile objects should make sure that the destructor only gets called when there are no other threads trying to access the ZipFile.